bionic/persistence.py

```diff
@@ -924,65 +928,17 @@ def valid_filename_from_query(query):
     return query.dnode.to_descriptor().replace(" ", "-")
 
 
-CACHE_SCHEMA_VERSION = 9
+CACHE_SCHEMA_VERSION = 10
 
 
 class YamlRecordParsingError(Exception):
     pass
 
 
-class ArtifactMetadataRecord:
-    """
-    Describes a persisted artifact.  Intended to be stored as a YAML file.
-    """
-
-    @classmethod
-    def from_content(cls, dnode, artifact_url, provenance, metadata_url, value_hash):
-        return cls(
-            body_dict=dict(
-                descriptor=dnode.to_descriptor(),
-                artifact_url=relativize_url(artifact_url, metadata_url),
-                provenance=provenance.to_dict(),
-                value_hash=value_hash,
-            )
-        )
-
-    @classmethod
-    def from_yaml(cls, yaml_str, metadata_url):
-        try:
-            body_dict = yaml.load(yaml_str, Loader=YamlLoader)
-        except yaml.error.YAMLError as e:
-            raise YamlRecordParsingError(f"Couldn't parse {cls.__name__}") from e
-        record = cls(body_dict=body_dict)
-        record.artifact_url = derelativize_url(record.artifact_url, metadata_url)
-        return record
-
-    def __init__(self, body_dict):
-        try:
-            self._dict = body_dict
-            self.descriptor = self._dict["descriptor"]
-            self.artifact_url = self._dict["artifact_url"]
-            self.provenance = Provenance.from_dict(self._dict["provenance"])
-            self.value_hash = self._dict["value_hash"]
-        except KeyError as e:
-            raise YamlRecordParsingError(
-                f"YAML for ArtifactMetadataRecord was missing field: {e}"
-            )
-
-    def to_yaml(self):
-        return yaml.dump(
-            self._dict,
-            default_flow_style=False,
-            encoding=None,
-            Dumper=YamlDumper,
-        )
-
-    def __repr__(self):
-        return f"ArtifactMetadataRecord({self.descriptor!r})"
-
-
-# TODO Could we use something like the cattr library instead of our own custom dict
-# stuff?
+# We don't use frozen=True because Provenance has list fields, which are not hashable.
```

Comment:
You're right, `frozen=True` will still work when the fields aren't hashable, although it will make `attrs` define a hash method for you. (The method will still fail, but with a slightly more confusing error message than the default.) I guess I think of `frozen` as suggesting that the object should be hashable.

Overall I don't have strong feeling about whether to use `frozen=True` here; WDYT?

Reply:
Ah, I see. I saw the documentation which said hash is only generated when both `frozen` and `eq` is set to `True`. What I didn't realize was that default `None` value for `eq` is I guess treated as `True`. 

I don't have strong feeling about it either. You can leave it as it is.